base/sort.jl

```diff
@@ -515,8 +515,12 @@ function sort!(v::AbstractVector, lo::Integer, hi::Integer, ::InsertionSortAlg,
     @inbounds for i = lo+1:hi
         j = i
         x = v[i]
-        while j > lo && lt(o, x, v[j-1])
-            v[j] = v[j-1]
+        while j > lo
+            y = v[j-1]
```

Comment:
Do we have any case where the compiler doesn't already do this optimization?
The only case I can think of would be stateful getindex
```julia
function getindex(x::LogAtReadArray, i::Int)
    println(i)
    getindex(x.data, i)
end
```
which isn't particularly compelling.

Reply:
any moderately complex or non-inlined getindex might not be eliminated

Reply:
I guess those likely exist, I've just never seen a subtype of `AbstractVector` with a complex `getindex`.

Reply:
any non-inlined `lt` user-method will also disallow this from occurring as a direct optimization (it would be semantically invalid)

Reply:
I don't see how that would be semantically invalid, but I'm not familiar with the relevant compiler semantics.

Reply:
It could throw for example.

Reply:
I see now! `lt` could mutate `v`. That would make this optimization change behavior. That behavior change is okay with me, but not okay with the compiler. The compiler cannot store the result of `v[j - 1]` unless `v[j - 1]` is pure with respect to the interleaving calls (including `lt`) which it can only know if both `getindex` and `lt` are simple and inlined.

I must have made a mistake in diffing. You're right, there is an extra load in the before case.

I'd be hard-pressed to find a case where this makes a measurable runtime difference, but I now see why it is an improvement in theory.